Kernel/Devices/SerialDevice.cpp

```diff
@@ -49,13 +49,13 @@ bool SerialDevice::can_write(const FileDescription&, size_t) const
     return (get_line_status() & EmptyTransmitterHoldingRegister) != 0;
 }
 
-KResultOr<size_t> SerialDevice::write(FileDescription&, u64, const UserOrKernelBuffer& buffer, size_t size)
+KResultOr<size_t> SerialDevice::write(FileDescription& description, u64, const UserOrKernelBuffer& buffer, size_t size)
 {
     if (!size)
         return 0;
 
-    if (!(get_line_status() & EmptyTransmitterHoldingRegister))
-        return 0;
+    if (!can_write(description, size))
+        return EAGAIN;
```

Comment:
Hmm, but isn't this still racy?

If we get preempted right after `can_write()` returns true, and someone else writes to this device, changing the state, we are now in the same place we didn't want to get to.